runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -659,83 +754,196 @@ private void setCurrentOutputWatermark(long currentOutputWatermark) {
-        @Override
-        public <T> void output(TupleTag<T> tag, WindowedValue<T> value) {
-          // with tagged outputs we can't get around this because we don't
-          // know our own output type...
-          @SuppressWarnings("unchecked")
-          WindowedValue<OutputT> castValue = (WindowedValue<OutputT>) value;
-          output.collect(new StreamRecord<>(castValue));
+  public static class BufferedOutputManager<OutputT> implements
+      DoFnRunners.OutputManager {
+
+    private final BiMap<TupleTag<?>, Integer> tagToLabel;
+    private TupleTag<OutputT> mainTag;
+    private Map<TupleTag<?>, OutputTag<WindowedValue<?>>> mapping;
+    protected Output<StreamRecord<WindowedValue<OutputT>>> output;
+
+    private boolean openBuffer = false;
+    private BagState<KV<Integer, WindowedValue<?>>> bufferState;
+
+    BufferedOutputManager(
+        Output<StreamRecord<WindowedValue<OutputT>>> output,
+        TupleTag<OutputT> mainTag,
+        Map<TupleTag<?>, OutputTag<WindowedValue<?>>> mapping,
+        final Map<TupleTag<?>, Coder<WindowedValue<?>>> coderMapping,
+        StateInternals stateInternals) {
+      this.output = output;
+      this.mainTag = mainTag;
+      this.mapping = mapping;
+
+      tagToLabel = transformTupleTagsToLabels(mainTag, mapping);
+
+      StateTag<BagState<KV<Integer, WindowedValue<?>>>> bufferTag =
+          StateTags.bag("bundle-buffer-tag", new TaggedKvCoder(tagToLabel, coderMapping));
+      bufferState = stateInternals.state(StateNamespaces.global(), bufferTag);
+    }
+
+    /**
+     * Transform TupleTags to labels. Integer is easier to serialize than TupleTag.
+     * Returns a BiMap for easier reversal query.
+     */
+    private static BiMap<TupleTag<?>, Integer> transformTupleTagsToLabels(
+        TupleTag<?> mainTag,
+        Map<TupleTag<?>, OutputTag<WindowedValue<?>>> mapping) {
+
+      BiMap<TupleTag<?>, Integer> tagToLabel = HashBiMap.create();
+      int count = 0;
+      tagToLabel.put(mainTag, count++);
+      for (TupleTag<?> key : mapping.keySet()) {
+        if (!tagToLabel.containsKey(key)) {
+          tagToLabel.put(key, count++);
         }
-      };
+      }
+      return tagToLabel;
+    }
+
+    void openBuffer() {
+      this.openBuffer = true;
+    }
+
+    void closeBuffer() {
+      this.openBuffer = false;
+    }
+
+    @Override
+    public <T> void output(TupleTag<T> tag, WindowedValue<T> value) {
+      if (!openBuffer) {
+        emit(tag, value);
+      } else {
+        bufferState.add(KV.<Integer, WindowedValue<?>>of(tagToLabel.get(tag), value));
+      }
+    }
+
+    /**
+     * Flush elements of bufferState to Flink Output. This method can't be invoke in
+     * {@link #snapshotState(StateSnapshotContext)}
+     */
+    void flushBuffer() {
+      for (KV<Integer, WindowedValue<?>> tuple2 : bufferState.read()) {
+        emit(tagToLabel.inverse().get(tuple2.getKey()), (WindowedValue) tuple2.getValue());
```

Comment:
`HashBiMap.inverse()` just provides an interface that does not have any substantial performance loss.